airflow/jobs.py

```diff
@@ -2372,20 +2389,33 @@ def _execute(self):
         executor = self.executor
         executor.start()
 
-        self._execute_for_run_dates(run_dates=run_dates,
-                                    ti_status=ti_status,
-                                    executor=executor,
-                                    pickle_id=pickle_id,
-                                    start_date=start_date,
-                                    session=session)
-
-        executor.end()
-        session.commit()
-        session.close()
+        ti_status.total_runs = len(run_dates)  # total dag runs in backfill
 
-        err = self._collect_errors(ti_status=ti_status, session=session)
-        if err:
-            raise AirflowException(err)
+        try:
+            while len(run_dates) > 0:
+                self._execute_for_run_dates(run_dates=run_dates,
+                                            ti_status=ti_status,
+                                            executor=executor,
+                                            pickle_id=pickle_id,
+                                            start_date=start_date,
+                                            session=session)
+
+                run_dates = [run_date for run_date in run_dates
```

Comment:
The `while` check with `any(run_date not in ti_status.executed_dag_run_dates for run_date in run_dates)` would not be enough, we'd still need to iterate the `run_dates` again to pass the remaining dates to `self._execute_for_run_dates`, otherwise we'll keep passing all the run_dates even the ones already processed - originally I mutated the same variable as a way to filter out the dates already processed and avoiding doing the same iteration of dates multiple times. 
I could replace the mutation of `run_dates` by a control `int` variable, effectively a counter, that keeps the number of remaining tasks and gets updated after processing the dates by doing: `len(run_dates) - len(ti_status.executed_dag_run_dates)`.